File: packages/sdk/go/tap/rfc9421.go
package tap

import (
	"crypto/ed25519"
	"crypto/rand"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"strings"
	"time"
)

type SignatureAlgorithm string

const (
	Ed25519       SignatureAlgorithm = "ed25519"
	RsaPssSha256  SignatureAlgorithm = "rsa-pss-sha256"
)

type SignatureParams struct {
	Created int64
	Expires int64
	KeyID   string
	Alg     SignatureAlgorithm
	Nonce   string
	Tag     string
}

type SignatureComponents struct {
	Authority string
	Path      string
}

type SignatureResult struct {
	SignatureInput string
	Signature      string
}

func CreateSignatureBase(components *SignatureComponents, params *SignatureParams) string {
	var lines []string

	lines = append(lines, fmt.Sprintf(`"@authority": %s`, components.Authority))
	lines = append(lines, fmt.Sprintf(`"@path": %s`, components.Path))

	signatureParamsValue := fmt.Sprintf(
		`("@authority" "@path"); created=%d; expires=%d; keyid="%s"; alg="%s"; nonce="%s"; tag="%s"`,
		params.Created,
		params.Expires,
		params.KeyID,
		params.Alg,
		params.Nonce,
		params.Tag,
	)

	lines = append(lines, fmt.Sprintf(`"@signature-params": %s`, signatureParamsValue))

	return strings.Join(lines, "\n")
}

func SignEd25519(
	components *SignatureComponents,
	params *SignatureParams,
	privateKey ed25519.PrivateKey,
) (*SignatureResult, error) {
	signatureBase := CreateSignatureBase(components, params)
	message := []byte(signatureBase)

	signature := ed25519.Sign(privateKey, message)
	signatureB64 := base64.StdEncoding.EncodeToString(signature)

	signatureInput := fmt.Sprintf(
		`sig2=("@authority" "@path"); created=%d; expires=%d; keyid="%s"; alg="%s"; nonce="%s"; tag="%s"`,
		params.Created,
		params.Expires,
		params.KeyID,
		params.Alg,
		params.Nonce,
		params.Tag,
	)

	return &SignatureResult{
		SignatureInput: signatureInput,
		Signature:      fmt.Sprintf("sig2=:%s:", signatureB64),
	}, nil
}

func GenerateNonce() (string, error) {
	nonce := make([]byte, 16)
	if _, err := rand.Read(nonce); err != nil {
		return "", err
	}
	return hex.EncodeToString(nonce), nil
}

func GenerateEd25519Keypair() (ed25519.PublicKey, ed25519.PrivateKey, error) {
	pub, priv, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		return nil, nil, err
	}
	return pub, priv, nil
}

func GetCurrentTimestamp() int64 {
	return time.Now().Unix()
}

func PrivateKeyToBase64(privateKey ed25519.PrivateKey) string {
	return base64.StdEncoding.EncodeToString(privateKey.Seed())
}

func PrivateKeyFromBase64(b64Str string) (ed25519.PrivateKey, error) {
	seed, err := base64.StdEncoding.DecodeString(b64Str)
	if err != nil {
		return nil, err
	}
	return ed25519.NewKeyFromSeed(seed), nil
}

func PublicKeyToBase64(publicKey ed25519.PublicKey) string {
	return base64.StdEncoding.EncodeToString(publicKey)
}

File: packages/sdk/rust/src/tap/rfc9421.rs
use ed25519_dalek::{Keypair, PublicKey, SecretKey, Signature, Signer};
use rand::rngs::OsRng;
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Debug, Clone)]
pub enum SignatureAlgorithm {
    Ed25519,
    RsaPssSha256,
}

impl SignatureAlgorithm {
    pub fn as_str(&self) -> &str {
        match self {
            SignatureAlgorithm::Ed25519 => "ed25519",
            SignatureAlgorithm::RsaPssSha256 => "rsa-pss-sha256",
        }
    }
}

#[derive(Debug, Clone)]
pub struct SignatureParams {
    pub created: u64,
    pub expires: u64,
    pub key_id: String,
    pub alg: SignatureAlgorithm,
    pub nonce: String,
    pub tag: String,
}

#[derive(Debug, Clone)]
pub struct SignatureComponents {
    pub authority: String,
    pub path: String,
}

#[derive(Debug, Clone)]
pub struct SignatureResult {
    pub signature_input: String,
    pub signature: String,
}

pub struct RFC9421Signature;

impl RFC9421Signature {
    pub fn create_signature_base(
        components: &SignatureComponents,
        params: &SignatureParams,
    ) -> String {
        let mut lines = Vec::new();

        lines.push(format!(r#""@authority": {}"#, components.authority));
        lines.push(format!(r#""@path": {}"#, components.path));

        let signature_params_value = format!(
            r#"("@authority" "@path"); created={}; expires={}; keyid="{}"; alg="{}"; nonce="{}"; tag="{}""#,
            params.created,
            params.expires,
            params.key_id,
            params.alg.as_str(),
            params.nonce,
            params.tag
        );

        lines.push(format!(r#""@signature-params": {}"#, signature_params_value));

        lines.join("\n")
    }

    pub fn sign_ed25519(
        components: &SignatureComponents,
        params: &SignatureParams,
        keypair: &Keypair,
    ) -> SignatureResult {
        let signature_base = Self::create_signature_base(components, params);
        let message = signature_base.as_bytes();

        let signature: Signature = keypair.sign(message);
        let signature_b64 = base64::encode(signature.to_bytes());

        let signature_input = format!(
            r#"sig2=("@authority" "@path"); created={}; expires={}; keyid="{}"; alg="{}"; nonce="{}"; tag="{}""#,
            params.created,
            params.expires,
            params.key_id,
            params.alg.as_str(),
            params.nonce,
            params.tag
        );

        SignatureResult {
            signature_input,
            signature: format!("sig2=:{}:", signature_b64),
        }
    }

    pub fn generate_nonce() -> String {
        use rand::RngCore;
        let mut nonce = [0u8; 16];
        OsRng.fill_bytes(&mut nonce);
        hex::encode(nonce)
    }

    pub fn generate_ed25519_keypair() -> Keypair {
        Keypair::generate(&mut OsRng)
    }

    pub fn get_current_timestamp() -> u64 {
        SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs()
    }
}
